File: server/src/routes/gifts.js
const { Router } = require('express');
const pool = require('../db/pool');

const router = Router();

const MAX_ITEMS_PER_GIFT = 3;
const MAX_GIFTS_PER_DAY = 5;
const GIFT_EXPIRY_DAYS = 7;

// POST /gifts/send
router.post('/send', async (req, res) => {
  const { toUid, items } = req.body;

  if (!toUid) {
    return res.status(400).json({ error: 'toUid is required' });
  }
  if (!Array.isArray(items) || items.length === 0) {
    return res.status(400).json({ error: 'items must be a non-empty array' });
  }
  if (items.length > MAX_ITEMS_PER_GIFT) {
    return res.status(400).json({ error: `Max ${MAX_ITEMS_PER_GIFT} items per gift` });
  }
  if (toUid === req.user.uid) {
    return res.status(400).json({ error: 'Cannot send gift to yourself' });
  }

  try {
    // Verify friendship
    const friendship = await pool.query(
      'SELECT 1 FROM friends WHERE player_uid = $1 AND friend_uid = $2',
      [req.user.uid, toUid]
    );
    if (friendship.rows.length === 0) {
      return res.status(403).json({ error: 'Not friends with this player' });
    }

    // Check daily gift limit
    const todayGifts = await pool.query(
      `SELECT COUNT(*) FROM gifts
       WHERE from_uid = $1 AND to_uid = $2
         AND created_at >= NOW() - INTERVAL '1 day'`,
      [req.user.uid, toUid]
    );
    if (parseInt(todayGifts.rows[0].count) >= MAX_GIFTS_PER_DAY) {
      return res.status(400).json({ error: `Max ${MAX_GIFTS_PER_DAY} gifts per day to same player` });
    }

    const result = await pool.query(
      `INSERT INTO gifts (from_uid, to_uid, items)
       VALUES ($1, $2, $3)
       RETURNING id, created_at`,
      [req.user.uid, toUid, JSON.stringify(items)]
    );

    res.status(201).json({
      giftId: result.rows[0].id,
      createdAt: result.rows[0].created_at
    });
  } catch (err) {
    res.status(500).json({ error: 'Failed to send gift' });
  }
});

// GET /gifts — pending gifts for authenticated user
router.get('/', async (req, res) => {
  try {
    const result = await pool.query(
      `SELECT g.id, g.from_uid, p.display_name AS from_display_name, g.items, g.created_at
       FROM gifts g
       JOIN players p ON p.uid = g.from_uid
       WHERE g.to_uid = $1
         AND g.status = 'pending'
         AND g.created_at >= NOW() - INTERVAL '7 days'
       ORDER BY g.created_at DESC`,
      [req.user.uid]
    );
    res.json({ gifts: result.rows });
  } catch (err) {
    res.status(500).json({ error: 'Failed to fetch gifts' });
  }
});

// POST /gifts/claim/:giftId
router.post('/claim/:giftId', async (req, res) => {
  const { giftId } = req.params;

  try {
    const result = await pool.query(
      `UPDATE gifts SET status = 'claimed', claimed_at = NOW()
       WHERE id = $1 AND to_uid = $2 AND status = 'pending'
       RETURNING items`,
      [giftId, req.user.uid]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Gift not found or already claimed' });
    }

    res.json({ items: result.rows[0].items });
  } catch (err) {
    res.status(500).json({ error: 'Failed to claim gift' });
  }
});

module.exports = router;

File: server/lib/camp_fire/gifts.ex
defmodule CampFire.Gifts do
  import Ecto.Query
  alias CampFire.Repo
  alias CampFire.Gifts.Gift
  alias CampFire.Accounts.Player

  @max_items_per_gift 3
  @max_gifts_per_day 5
  @gift_expiry_days 7

  def send_gift(from_uid, to_uid, items) do
    cond do
      to_uid == from_uid ->
        {:error, "Cannot send gift to yourself"}

      not is_list(items) or length(items) == 0 ->
        {:error, "items must be a non-empty array"}

      length(items) > @max_items_per_gift ->
        {:error, "Max #{@max_items_per_gift} items per gift"}

      gifts_today(from_uid, to_uid) >= @max_gifts_per_day ->
        {:error, "Max #{@max_gifts_per_day} gifts per day to same player"}

      true ->
        %Gift{}
        |> Gift.changeset(%{from_uid: from_uid, to_uid: to_uid, items: items})
        |> Repo.insert()
    end
  end

  def pending_gifts(to_uid) do
    cutoff = DateTime.add(DateTime.utc_now(), -@gift_expiry_days * 86400, :second)

    from(g in Gift,
      join: p in Player, on: p.uid == g.from_uid,
      where: g.to_uid == ^to_uid and g.status == "pending" and g.inserted_at >= ^cutoff,
      order_by: [desc: g.inserted_at],
      select: %{id: g.id, from_uid: g.from_uid, from_name: p.display_name, items: g.items, created_at: g.inserted_at}
    )
    |> Repo.all()
  end

  def claim_gift(gift_id, to_uid) do
    query =
      from(g in Gift,
        where: g.id == ^gift_id and g.to_uid == ^to_uid and g.status == "pending"
      )

    case Repo.one(query) do
      nil ->
        {:error, :not_found}

      gift ->
        gift
        |> Gift.changeset(%{status: "claimed", claimed_at: DateTime.utc_now() |> DateTime.truncate(:second)})
        |> Repo.update()
        |> case do
          {:ok, claimed} -> {:ok, claimed.items}
          {:error, _} -> {:error, :update_failed}
        end
    end
  end

  defp gifts_today(from_uid, to_uid) do
    cutoff = DateTime.add(DateTime.utc_now(), -86400, :second)

    from(g in Gift,
      where: g.from_uid == ^from_uid and g.to_uid == ^to_uid and g.inserted_at >= ^cutoff
    )
    |> Repo.aggregate(:count)
  end
end
